File: Hypedsearchservice/proteinmatch.fs
module proteinmatch

open System.IO
open System.Net
open Microsoft.Extensions.Logging
open Microsoft.Azure.Functions.Worker.Http
open Microsoft.Azure.Functions.Worker
open Newtonsoft.Json

type ProteinMatch = {ProteinName:string; Weight:float; StartIndex:int; EndIndex:int}
type WeightProteinMatches = {Weight:float; ProteinMatchs:ProteinMatch seq}
type AllWeightProteinMatches = WeightProteinMatches seq

let getWeightProteinMatch weight kmerLength (context: FunctionContext) =
    let fileName = "all_weight_protein_matches_kmer" + kmerLength.ToString() + ".json"
    let assemblyPath = context.FunctionDefinition.PathToAssembly
    let directoryPath = assemblyPath.Substring(0, assemblyPath.LastIndexOf('\\')).TrimEnd()
    let filePath = System.IO.Path.Combine(directoryPath, fileName)
    let json = File.ReadAllText(filePath)
    let contents = JsonConvert.DeserializeObject<AllWeightProteinMatches> json
    let found = contents|> Seq.filter(fun x -> x.Weight = weight)
    JsonConvert.SerializeObject found

[<Function "proteinmatch">]
let run ([<HttpTrigger(AuthorizationLevel.Anonymous, "GET")>] req: HttpRequestData, executionContext: FunctionContext) =
    let queryString = req.Url.Query
    let tokens = System.Web.HttpUtility.ParseQueryString(queryString)
    let kmerLength = tokens.Get("kmer_length")
    let weight = System.Double.Parse(tokens.Get("weight"))
    let res = req.CreateResponse(HttpStatusCode.OK)
    res.Headers.Add("Content-Type", "text/plain; charset=utf-8")
    let response = getWeightProteinMatch weight kmerLength executionContext
    res.WriteString(response);
    res

File: Hypedsearchservice/hypedsearchservice.api/proteinmatch.cs
using System;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.AspNetCore.Http;
using Newtonsoft.Json;
using hypedsearchservice.core;


namespace hypedsearchservice
{

    public static class proteinmatch
    {
        //http://hypedsearchservice.azurewebsites.net/api/proteinmatch?ion_charge=B&weight=218.07&ppm_tolerance=.01
        [FunctionName("proteinmatch")]
        public static async Task<IActionResult> Run([HttpTrigger(AuthorizationLevel.Anonymous, "get")] HttpRequest req, ExecutionContext context)
        {
            try
            {
                var ion_charge = req.Query["ion_charge"];
                var weight = Double.Parse(req.Query["weight"]);
                var ppm_tolerance = Double.Parse(req.Query["ppm_tolerance"]);
                var matcher = new ProteinMatcher();
                var proteinMatches = matcher.GetProteinMatchesViaSingle(ion_charge, weight, ppm_tolerance);
                var serializeProteinMatches = JsonConvert.SerializeObject(proteinMatches);
                return new OkObjectResult(serializeProteinMatches);
            }
            catch(Exception e)
            {
                return new ObjectResult(e.ToString());
            }

        }
    }
}

